doc/user: Add metrics and service-monitor docs (#719)
operator-framework/operator-sdk: doc/user/metrics/README.md

# Monitoring with Prometheus

[Prometheus][prometheus] is an open-source systems monitoring and alerting toolkit. Below is the overview of the different helpers that exist in operator-sdk to help setup metrics in the generated operator.

## Metrics in operator-sdk

> The `func ExposeMetricsPort(ctx context.Context, port int32) (*v1.Service, error)` is a helper function that exposes general metrics about the running program. These metrics are inherited from controller-runtime. The metrics are by default served on `:8383/metrics`. A [Service][service] object is created with the metrics port exposed, which can be then accessed by Prometheus. The Service object is [garbage collected][gc] by the owner, which is determined based on the pods owner in which the leader is running in.

Reviewer: ```suggestion
The `func ExposeMetricsPort(ctx context.Context, port int32) (*v1.Service, error)` function exposes general metrics about the running program. These metrics are inherited from controller-runtime. This helper function creates a [Service][service] object with the metrics port exposed, which can then be accessed by Prometheus. The Service object is [garbage collected][gc] when the leader pod's root owner is deleted.
```

Do you think we need an example for the "leader pod's root owner", or whatever wording we end up with on that?

Author: An example sure, what were you thinking of?

> To modify the port the metrics are exposed on, change the variable `metricsHost string = ":8383"` variable in the `cmd/manager/main.go` file of the generated operator.

Reviewer: I would suggest moving the sentence in the above paragraph to the first sentence of this paragraph.

```suggestion
By default, the metrics are served on `0.0.0.0:8383/metrics`. To modify the port the metrics are exposed on, change the `var metricsPort int32 = 8383` variable in the `cmd/manager/main.go` file of the generated operator.
```

Should we mention anything about changing (or not changing) the `metricsHost` variable?

Author: > I would suggest moving the sentence in the above paragraph to the first sentence of this paragraph.

Sorry, not sure I understand that. :)

### Usage:

```go
    import(
     "github.com/operator-framework/operator-sdk/pkg/metrics"
	"sigs.k8s.io/controller-runtime/pkg/manager"
    )

    // Change the below variables to serve metrics on different host or port.
    var metricsHost = "0.0.0.0"
    var metricsPort int32 = 8383

    // Pass metrics address to controller-runtime manager
	mgr, err := manager.New(cfg, manager.Options{
		Namespace:          namespace,
		MetricsBindAddress: fmt.Sprintf("%s:%d", metricsHost, metricsPort),
	})
	
    ...

	// Create Service object to expose the metrics port.
	_, err = metrics.ExposeMetricsPort(ctx, metricsPort)
	if err != nil {
        // handle error
		log.Info(err.Error())

> 	}

Author: I did the single dots with no comments in case someone copies this so they know it's meant to be added, comments would be ignored, hope that makes sense.

# Monitoring with Prometheus

[Prometheus][prometheus] is an open-source systems monitoring and alerting toolkit. Below is the overview of the different helpers that exist in operator-sdk to help setup metrics in the generated operator.

## Metrics in operator-sdk

The `func ExposeMetricsPort(ctx context.Context, port int32) (*v1.Service, error)` function exposes general metrics about the running program. These metrics are inherited from controller-runtime. This helper function creates a [Service][service] object with the metrics port exposed, which can then be accessed by Prometheus. The Service object is [garbage collected][gc] when the leader pod's root owner is deleted.

By default, the metrics are served on `0.0.0.0:8383/metrics`. To modify the port the metrics are exposed on, change the `var metricsPort int32 = 8383` variable in the `cmd/manager/main.go` file of the generated operator.

### Usage:

```go
    import(
        "github.com/operator-framework/operator-sdk/pkg/metrics"

> 	    "sigs.k8s.io/controller-runtime/pkg/manager"

Reviewer 2: Fix indentation to align with the other import:
```suggestion
        "sigs.k8s.io/controller-runtime/pkg/manager"
```

    )

    func main() {

        ...

        // Change the below variables to serve metrics on different host or port.
        var metricsHost = "0.0.0.0"
        var metricsPort int32 = 8383

        // Pass metrics address to controller-runtime manager
        mgr, err := manager.New(cfg, manager.Options{
            Namespace:          namespace,
            MetricsBindAddress: fmt.Sprintf("%s:%d", metricsHost, metricsPort),
        })
        
        ...

        // Create Service object to expose the metrics port.
        _, err = metrics.ExposeMetricsPort(ctx, metricsPort)
        if err != nil {
            // handle error
            log.Info(err.Error())
        }

        ...
    }
```

> Note: The above example is already present in `cmd/manager/main.go` in all the operators generated with operator-sdk.

Reviewer 2: Nit: Just to highlight the note.
```suggestion
*Note:* The above example is already present in `cmd/manager/main.go` in all the operators generated with operator-sdk.
```
